src/client/browserType.ts

```diff
@@ -144,61 +146,48 @@ export class BrowserType extends ChannelOwner<channels.BrowserTypeChannel, chann
           try {
             await (params as any).__testHookBeforeCreateBrowser();
           } catch (e) {
             reject(e);
           }
         }
-        const prematureCloseListener = (params: { error?: channels.SerializedError }) => {
-          reject(new Error(`WebSocket server disconnected ${params.error!.error?.message}`));
-        };
-        pipe.on('closed', prematureCloseListener);
-
-        pipe.on('opened', async () => {
-          const playwright = await connection!.initializePlaywright();
-
-          if (!playwright._initializer.preLaunchedBrowser) {
-            reject(new Error('Malformed endpoint. Did you use launchServer method?'));
-            pipe.close().catch(() => {});
-            return;
-          }
-
-          const browser = Browser.from(playwright._initializer.preLaunchedBrowser!);
-          browser._logger = logger;
-          browser._remoteType = 'owns-connection';
-          browser._setBrowserType((playwright as any)[browser._name]);
-          const closeListener = (param: { error?: Error }) => {
-            // Emulate all pages, contexts and the browser closing upon disconnect.
-            for (const context of browser.contexts()) {
-              for (const page of context.pages())
-                page._onClose();
-              context._onClose();
-            }
-            browser._didClose();
-            connection.didDisconnect(kBrowserClosedError);
-            if (param.error)
-              reject(new Error(param.error + '. Most likely ws endpoint is incorrect'));
-          };
-          pipe.off('closed', prematureCloseListener);
-          pipe.on('closed', closeListener);
-          browser.on(Events.Browser.Disconnected, () => {
-            playwright._cleanup();
-            pipe.off('closed', closeListener);
-            pipe.close().catch(() => {});
-          });
-          fulfill(browser);
+        const playwright = await connection!.initializePlaywright();
+        if (!playwright._initializer.preLaunchedBrowser) {
+          reject(new Error('Malformed endpoint. Did you use launchServer method?'));
+          pipe.close().catch(() => {});
+          return;
+        }
+        browser = Browser.from(playwright._initializer.preLaunchedBrowser!);
+        browser._logger = logger;
+        browser._remoteType = 'owns-connection';
+        browser._setBrowserType((playwright as any)[browser._name]);
+        browser.on(Events.Browser.Disconnected, () => {
+          playwright._cleanup();
+          pipe.close().catch(() => {});
         });
+        fulfill(browser);
       });
 
-      try {
-        return await Promise.race([successPromise, timeoutPromise]);
-      } finally {
-        if (timer)
-          clearTimeout(timer);
-      }
+      const result = await raceAgainstDeadline(createBrowserPromise, deadline);
+      if (result.result)
+        return result.result;
+      else
+        throw new Error(`Timeout ${params.timeout}ms exceeded`);
```

Comment:
Let's ensure the pipe and the browser get closed if it just took longer to connect.

Reply:
Done.